src/storages/request_queue.js

```diff
@@ -404,19 +405,24 @@ export class RequestQueue {
         this._cacheRequest(getRequestId(request.uniqueKey), queueOperationInfo);
         queueOperationInfo.request = request;
 
+        // Previously we were loosing the async context here by calling `setTimeout()` with
+        // async callback and returning early, before the actual reclaim happened. To fix that
+        // we now use promised `setTimeout()` and await it, so once `reclaimRequest` resolves,
+        // we now the request was actually reclaimed.
+
         // Wait a little to increase a chance that the next call to fetchNextRequest() will return the request with updated data.
         // This is to compensate for the limitation of DynamoDB, where writes might not be immediately visible to subsequent reads.
-        setTimeout(() => {
-            if (!this.inProgress.has(request.id)) {
-                this.log.warning('The request is no longer marked as in progress in the queue?!', { requestId: request.id });
-                return;
-            }
+        await setTimeout(STORAGE_CONSISTENCY_DELAY_MILLIS);
```

Comment:
This is executed in each call of `_fetchNextRequest()` no? Which means all fetches will now take 3+ seconds. That's not acceptable.

Reply:
Yes, but the call itself is done from a separate task that does not block other tasks, or not? I thought the wait will be only when request fails and only in the process that handles such request.

We need to wait, this was the exact cause of the zero concurrency bug. We can't just postpone async calls via `setTimeout` and ignore their resolution. I see that pattern all over our repositories and I am more and more confident it is just wrong...

Reply:
> Yes, but the call itself is done from a separate task that does not block other tasks, or not? I thought the wait will be only when request fails and only in the process that handles such request.

Verified locally, if I have 5 requests in the queue, and add the pre nav timeout only to the first one, it only stalls the first one. The processing takes 1:20min with the timeouts commented out, and 1:30min with awaiting them as in the PR (so roughly plus 3 seconds times 3 retries). 

```ts
const requestQueue = await Apify.openRequestQueue();
await requestQueue.addRequest({ url: 'http://www.example.com/?q=1' });
await requestQueue.addRequest({ url: 'http://www.example.com/?q=2' });
await requestQueue.addRequest({ url: 'http://www.example.com/?q=3' });
await requestQueue.addRequest({ url: 'http://www.example.com/?q=4' });
await requestQueue.addRequest({ url: 'http://www.example.com/?q=5' });
const handlePageFunction = jest.fn();

const crawler = new Apify.PuppeteerCrawler({
    requestQueue,
    handlePageTimeoutSecs: 5,
    navigationTimeoutSecs: 5,
    preNavigationHooks: [
        async ({ request }) => {
            console.log('pre nav', request.url, new Date());
            if (request.url.endsWith('/?q=1')) {
                console.log('wait', request.url);
                await Apify.utils.sleep(20e3);
            }
        },
    ],
    handlePageFunction,
});
```

and the logs:

```
pre nav http://www.example.com/?q=1 2021-11-26T12:48:18.906Z
wait http://www.example.com/?q=1
pre nav http://www.example.com/?q=2 2021-11-26T12:48:19.035Z
pre nav http://www.example.com/?q=3 2021-11-26T12:48:19.437Z
pre nav http://www.example.com/?q=4 2021-11-26T12:48:19.668Z
pre nav http://www.example.com/?q=5 2021-11-26T12:48:19.880Z

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=1","retryCount":1,"id":"7q1bruYFfePreNI"}
Error: handleRequestFunction timed out after 20 seconds.

pre nav http://www.example.com/?q=1 2021-11-26T12:48:41.586Z
wait http://www.example.com/?q=1

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=1","retryCount":2,"id":"7q1bruYFfePreNI"}
Error: handleRequestFunction timed out after 20 seconds.

pre nav http://www.example.com/?q=1 2021-11-26T12:49:04.590Z
wait http://www.example.com/?q=1

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=1","retryCount":3,"id":"7q1bruYFfePreNI"}
Error: handleRequestFunction timed out after 20 seconds.

pre nav http://www.example.com/?q=1 2021-11-26T12:49:27.644Z
wait http://www.example.com/?q=1

ERROR PuppeteerCrawler: Request failed and reached maximum retries {"id":"7q1bruYFfePreNI","url":"http://www.example.com/?q=1","method":"GET","uniqueKey":"http://www.example.com?q=1"}
Error: handleRequestFunction timed out after 20 seconds.
```

If I stall this way 2 requests (tried with first and third), we get the exact same 1:30 min time, as those things work in parallel. 

```
pre nav http://www.example.com/?q=1 2021-11-26T12:58:10.221Z
wait http://www.example.com/?q=1
pre nav http://www.example.com/?q=2 2021-11-26T12:58:10.381Z
pre nav http://www.example.com/?q=3 2021-11-26T12:58:10.748Z
wait http://www.example.com/?q=3
pre nav http://www.example.com/?q=4 2021-11-26T12:58:19.924Z
pre nav http://www.example.com/?q=5 2021-11-26T12:58:20.164Z

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=1","retryCount":1,"id":"7q1bruYFfePreNI"}
Error: handleRequestFunction timed out after 20 seconds.

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=3","retryCount":1,"id":"Qz0fpc27goVAjRi"}
Error: handleRequestFunction timed out after 20 seconds.

pre nav http://www.example.com/?q=1 2021-11-26T12:58:32.886Z
wait http://www.example.com/?q=1
pre nav http://www.example.com/?q=3 2021-11-26T12:58:33.786Z
wait http://www.example.com/?q=3

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=1","retryCount":2,"id":"7q1bruYFfePreNI"}
Error: handleRequestFunction timed out after 20 seconds.

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=3","retryCount":2,"id":"Qz0fpc27goVAjRi"}
Error: handleRequestFunction timed out after 20 seconds.

pre nav http://www.example.com/?q=1 2021-11-26T12:58:55.896Z
wait http://www.example.com/?q=1
pre nav http://www.example.com/?q=3 2021-11-26T12:58:56.765Z
wait http://www.example.com/?q=3

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=1","retryCount":3,"id":"7q1bruYFfePreNI"}
Error: handleRequestFunction timed out after 20 seconds.

ERROR PuppeteerCrawler: handleRequestFunction failed, reclaiming failed request back to the list or queue {"url":"http://www.example.com/?q=3","retryCount":3,"id":"Qz0fpc27goVAjRi"}
Error: handleRequestFunction timed out after 20 seconds.

pre nav http://www.example.com/?q=1 2021-11-26T12:59:18.948Z
wait http://www.example.com/?q=1
pre nav http://www.example.com/?q=3 2021-11-26T12:59:19.790Z
wait http://www.example.com/?q=3

ERROR PuppeteerCrawler: Request failed and reached maximum retries {"id":"7q1bruYFfePreNI","url":"http://www.example.com/?q=1","method":"GET","uniqueKey":"http://www.example.com?q=1"}
Error: handleRequestFunction timed out after 20 seconds.

ERROR PuppeteerCrawler: Request failed and reached maximum retries {"id":"Qz0fpc27goVAjRi","url":"http://www.example.com/?q=3","method":"GET","uniqueKey":"http://www.example.com?q=3"}
Error: handleRequestFunction timed out after 20 seconds.
```

To me this can add at max 10s to total run time, in case one of the very last requests would fail. Sounds more acceptable? :]